text/3064-bitfield.md

```diff
@@ -0,0 +1,384 @@
+struct X {
+    unsigned char a: 1;
+    unsigned char b: 8;
+    unsigned char  : 0;
+    unsigned char d: 2;
+};
+```
+
+As you can see in the example, unnamed bit-fields in C are written with the
+reserved identifier `_` in Rust. When you have a struct with an unnamed
+bit-field, you cannot access that field. On the other hand, you also do not have
+to define a value for that field when constructing the struct:
+
+```rust
+let x = X { a: 1, b: 1, d: 1 };
+let X { a, b, d } = x;
+```
+
+Just like in C, you cannot take the address of a bit-field:
+
+```rust
+fn f(x: &X) -> &u8 {
+    &x.a // !!! does not compile
+}
+```
+
+The value of the `N` in `bitfield(N)` must be a non-negative integer literal.
+This literal must not be larger than the bit-size of the type of the field.
+
+In debug builds, when you write to a bit-field, Rust performs overflow checks.
+If the value to be stored does not fit in the bit-field, the overflow check
+fails and the operation panics:
+
+```rust
+fn f(x: &mut X) {
+    x.a = 6; // !!! overflow check fails
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The attribute `bitfield(N)` is added. `N` must be an integer literal. `N` is
+called the *width* of the bit-field. This attribute can only be applied to fields
+of a `repr(C)` struct. Such a field is called a *bit-field*. A bit-field must
+have one of the types
+
+- `bool`,
+- `u8`, `u16`, `u32`, `u64`, `u128`, `usize`,
+- `i8`, `i16`, `i32`, `i64`, `i128`, `isize`
+
+modulo type aliases.
+
+The width of a bit-field with type `T` must be in the range `[0,
+bit_width_of(T)]`. For the integer types, `bit_width_of` is defined as `8 *
+sizeof`. For `bool`, `bit_width_of` is 1.
+
+A bit-field can have the name `_`. Such fields are only used to modify the layout
+of the struct. Consider the example
+
+```rust
+#[repr(C)]
+struct X {
+    a: u32,
+    #[bitfield(6)] _: u32,
+    #[bitfield(1)] b: u32,
+    #[bitfield(5)] _: u32,
+    #[bitfield(1)] c: u32,
+}
+```
+
+This struct behaves like
+
+```rust
+#[repr(C)]
+#[insert_unnamed_bitfield_after(0, 6, u32)]
+#[insert_unnamed_bitfield_after(1, 5, u32)]
+struct X {
+    a: u32,
+    #[bitfield(1)] b: u32,
+    #[bitfield(1)] c: u32,
+}
+```
+
+if the attribute `insert_unnamed_bitfield_after` existed and caused the layout
+to be modified in the same way as by the unnamed bit-field as described below.
+
+In particular, such a field cannot be accessed, need not and cannot appear in
+the construction of such a struct, cannot appear in the deconstruction of such a
+struct.
+
+Nevertheless, such fields appear like regular fields in the rustdoc output if
+they have the `pub` visibility.
+
+If the width of the bit-field is `0`, then the name of the field must be `_`.
+
+A field named `_` is called an *unnamed* field. All other fields are called
+*named* fields. Each named field annotated with `bitfield(N)` occupies `N` bits
+of storage. Unnamed fields do not occupy any storage.
```

Comment:
I agree that this needs clarification.

```suggestion
of storage. Unnamed bit-fields act like padding within the structure; they affect the positioning of other fields, but are not themselves accessible. They act in every way like padding bits.
```

Reply:
I'm wary of the "in every way like padding bits" phrasing because it brings to mind undefined padding bytes (a la [this comment](https://github.com/rust-lang/rfcs/pull/3064#discussion_r556115242))

Reply:
I've rewritten the reference section to make it clear that `_` bit-fields are not fields.